skbeam/core/correlation.py

```diff
@@ -911,93 +909,130 @@ def __init__(self, shape, mask=None, normalization=None,
                     'regular' : divide by pixel number
                     'symavg' : use symmetric averaging
                 Defaults to ['regular'] normalization
-
-            wrap : bool, optional
-                If False, assume dimensions don't wrap around. If True
-                    assume they do. The latter is useful for circular
-                    dimensions such as angle.
         '''
         if normalization is None:
             normalization = ['regular']
         elif not isinstance(normalization, list):
             normalization = list([normalization])
-
-        self.wrap = wrap
         self.normalization = normalization
 
         if mask is None:
             mask = np.ones(shape)
 
-        # the IDs for the image, called mask
-        self.mask = mask
-        # initialize all the masks for the correlation
+        self.shape = shape
+        self.ndim = len(shape)
+        if self.ndim == 1:
+            mask = mask.reshape((1, shape[0]))
+
+        if self.ndim > 2:
+            raise ValueError("Dimensions other than 1 or 2 not supported")
+
+        # A terse nomenclature was chosen for ease of reading code.
+        # Here is the Nomenclature (for both init and call sections):
+        #   1. pi, pj: the list of yindex (rows), xindex (cols) into original
+        #       image
+        #   2. pis, pjs : a subselection of the above
+        #   3. ppii, ppjj : the list of yindex (rows), xindex (cols) into
+        #       smaller sub images
+        #   4. ppiis, ppjjs : a subselection of the above
+        #   5. bind : the bin id per pixel (defined by mask)
+        #   6. idpos : a list of pointers into the position arrays for each id
+
+        # initialize subregions information for the correlation
+        # first find indices of subregions and sort them by subregion id
+        pi, pj = np.where(mask != 0)
+        # typing not necessary, but it may be better to be explicit
+        # (especially if we cythonize in the future)
+        bind = mask[pi, pj].astype(np.uint32)
+
+        # sort them so that we can easily index into subregions with one number
+        order = np.argsort(bind)
```

Comment:
you can credit Mark Sutton for that (he doesn't have a github). that's the integral piece he's basically added here, yeah pretty clever :-)